File: examples/bun/example02_enumerate_adapters.ts
import {
  DXBridgeLibrary,
  formatBytes,
  getStringArg,
  parseArgs,
  parseBackend,
} from "./dxbridge";

function printUsage(): void {
  console.log("Usage: bun run examples/bun/example02_enumerate_adapters.ts [--dll PATH] [--backend auto|dx11|dx12]");
}

function main(): void {
  const args = parseArgs(Bun.argv.slice(2));
  if (args.values.help === true) {
    printUsage();
    return;
  }

  const backend = parseBackend(getStringArg(args, "backend", "auto"));
  const dllPath = getStringArg(args, "dll");
  const dxbridge = new DXBridgeLibrary(dllPath || undefined);

  console.log(`Loaded: ${dxbridge.path}`);

  try {
    dxbridge.init(backend);
    const adapters = dxbridge.enumerateAdapters();
    if (adapters.length === 0) {
      console.log("No adapters reported by DXBridge.");
      return;
    }

    console.log(`Found ${adapters.length} adapter(s):`);
    for (const adapter of adapters) {
      const kind = adapter.isSoftware ? "software" : "hardware";
      console.log(
        `- #${adapter.index}: ${adapter.description} [${kind}] | VRAM=${formatBytes(adapter.dedicatedVideoMem)} | Shared=${formatBytes(adapter.sharedSystemMem)}`,
      );
    }
  } finally {
    dxbridge.shutdown();
    dxbridge.close();
  }
}

main();

File: examples/python/example_02_enumerate_adapters.py
from __future__ import annotations

import argparse

from dxbridge_ctypes import BACKEND_NAMES, DXBridgeLibrary, format_bytes


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Enumerate DXBridge adapters using the two-call ABI pattern."
    )
    parser.add_argument("--dll", help="Optional explicit path to dxbridge.dll")
    parser.add_argument(
        "--backend",
        choices=sorted(BACKEND_NAMES),
        default="auto",
        help="Backend hint used during DXBridge_Init before enumeration",
    )
    args = parser.parse_args()

    dxbridge = DXBridgeLibrary(args.dll)
    print(f"Loaded: {dxbridge.path}")

    try:
        dxbridge.init(BACKEND_NAMES[args.backend])
        adapters = dxbridge.enumerate_adapters()
        if not adapters:
            print("No adapters reported by DXBridge.")
            return 0

        print(f"Found {len(adapters)} adapter(s):")
        for adapter in adapters:
            description = adapter.description.decode("utf-8", errors="replace").rstrip(
                "\x00"
            )
            adapter_kind = "software" if adapter.is_software else "hardware"
            print(
                f"- #{adapter.index}: {description} [{adapter_kind}] | "
                f"VRAM={format_bytes(adapter.dedicated_video_mem)} | "
                f"Shared={format_bytes(adapter.shared_system_mem)}"
            )
    finally:
        dxbridge.shutdown()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
